StreamFragmentWidthAdapter(x, y, earlyLast = true) widthOf(y) = 2 *widthOf(x) triggers assertion in whenIndexed()

```scala
// minimal case to reproduce bug for case StreamFragmentWidthAdapter(x, y = x * 2, earlyLast = true)
case class AxisUpsizer(dataWidthIn : Int, dataWidthOut: Int) extends Component {
  require(dataWidthOut == (2 * dataWidthIn), "dataWidthOut must be a factor 2 of dataWidthIn")
  require(dataWidthOut % dataWidthIn == 0, "dataWidthOut must be an integer multiple of dataWidthIn")

  val io = new Bundle {
    val sink = slave Stream(Fragment(Bits(dataWidthIn bits)))
    val source = master Stream(Fragment(Bits(dataWidthOut bits)))
  }
  StreamFragmentWidthAdapter(io.sink, io.source, earlyLast = true)
}
```
(or any factor 2 with earlyLast = true) errors out as follows:
```scala
[error] Exception in thread "main" java.lang.AssertionError: assertion failed
[error]         at scala.Predef$.assert(Predef.scala:156)
[error]         at spinal.core.package$.assert(core.scala:463)
[error]         at spinal.lib.whenIndexed$.apply(Utils.scala:1320)
[error]         at spinal.lib.StreamFragmentWidthAdapter$$anon$56$$anon$41$$anonfun$41.apply$mcV$sp(Stream.scala:1601)
[error]         at spinal.core.when$.apply(when.scala:90)
[error]         at spinal.lib.StreamFragmentWidthAdapter$$anon$56$$anon$41.<init>(Stream.scala:1600)
[error]         at spinal.lib.StreamFragmentWidthAdapter$$anon$56.<init>(Stream.scala:1562)
[error]         at spinal.lib.StreamFragmentWidthAdapter$.apply(Stream.scala:1541)
[error]         at corundum.AxisUpsizer.<init>(AxisUpsizer.scala:84)
[error]         at corundum.AxisUpsizerSim$$anonfun$main$1.apply(AxisUpsizerSim.scala:18)
[error]         at corundum.AxisUpsizerSim$$anonfun$main$1.apply(AxisUpsizerSim.scala:18)
[error]         at spinal.core.internals.PhaseCreateComponent$$anonfun$impl$66.apply(Phase.scala:2404)
[error]         at spinal.core.fiber.Engine$$anonfun$create$1.apply$mcV$sp(AsyncCtrl.scala:147)
[error]         at spinal.core.fiber.AsyncThread$$anonfun$1.apply$mcV$sp(AsyncThread.scala:59)
[error]         at spinal.core.fiber.EngineContext$$anonfun$newJvmThread$1.apply$mcV$sp(AsyncCtrl.scala:39)
```

Root cause analysis (also see Gitter) is that ```buffer``` stores one input word, whereas the counter counts in [0,1] range, thus the number of bits required for indexes differs in the whenIndexed(), triggering the assertion.

Fix could be to use relaxedWidth = (factor <= 2) in 

```scala
          when(input.fire) {
            whenIndexed(buffer.subdivideIn(inputWidth bits), counter, relaxedWidth = (factor <= 2)) {
              _ := input.fragment.asBits
            }
          }
          whenIndexed(data.subdivideIn(inputWidth bits).dropRight(1), counter, relaxedWidth = (factor <= 2)) {
            _ := input.fragment.asBits
          }
```

--- a/lib/src/main/scala/spinal/lib/Stream.scala
+++ b/lib/src/main/scala/spinal/lib/Stream.scala
@@ -1598,11 +1598,11 @@ object StreamFragmentWidthAdapter {
           for((bit, id) <- dataMask.asBools.zipWithIndex) bit := counter >= id
 
           when(input.fire) {
-            whenIndexed(buffer.subdivideIn(inputWidth bits), counter) {
+            whenIndexed(buffer.subdivideIn(inputWidth bits), counter, relaxedWidth = (factor <= 2)) {
               _ := input.fragment.asBits
             }
           }
-          whenIndexed(data.subdivideIn(inputWidth bits).dropRight(1), counter) {
+          whenIndexed(data.subdivideIn(inputWidth bits).dropRight(1), counter, relaxedWidth = (factor <= 2)) {
             _ := input.fragment.asBits
           }
         }

--- a/lib/src/main/scala/spinal/lib/Stream.scala
+++ b/lib/src/main/scala/spinal/lib/Stream.scala
@@ -1598,11 +1598,11 @@ object StreamFragmentWidthAdapter {
           for((bit, id) <- dataMask.asBools.zipWithIndex) bit := counter >= id
 
           when(input.fire) {
-            whenIndexed(buffer.subdivideIn(inputWidth bits), counter, relaxedWidth = (factor <= 2)) {
+            whenIndexed(buffer.subdivideIn(inputWidth bits), counter, relaxedWidth = true) {
               _ := input.fragment.asBits
             }
           }
-          whenIndexed(data.subdivideIn(inputWidth bits).dropRight(1), counter, relaxedWidth = (factor <= 2)) {
+          whenIndexed(data.subdivideIn(inputWidth bits).dropRight(1), counter, relaxedWidth = true) {
             _ := input.fragment.asBits
           }
         }

--- a/tester/src/test/scala/spinal/tester/scalatest/Issue963Tester.scala
+++ b/tester/src/test/scala/spinal/tester/scalatest/Issue963Tester.scala
@@ -31,8 +31,8 @@ class Issue963Tester extends AnyFunSuite {
   test("Minimal example for Issue963 which fails Spinal elaboration") {
     class Dut extends Component {
       val io = new Bundle {
-          val sink = slave Stream(Fragment(Bits(8 bits)))
-          val source = master Stream(Fragment(Bits(16 bits)))
+        val sink = slave Stream (Fragment(Bits(8 bits)))
+        val source = master Stream (Fragment(Bits(16 bits)))
       }
       val x = Stream(Fragment(Bits(16 bits)))
 
@@ -40,12 +40,14 @@ class Issue963Tester extends AnyFunSuite {
       io.source <-< x
     }
 
-    SimConfig.withConfig(SpinalConfig(defaultClockDomainFrequency = FixedFrequency(4000 Hz))).compile(new Component{
-      val issue = new Dut()
-    })
-   .doSim{dut =>
-      dut.clockDomain.forkStimulus(10)
-      dut.clockDomain.waitSampling()
-    }
+    SimConfig
+      .withConfig(SpinalConfig(defaultClockDomainFrequency = FixedFrequency(4000 Hz)))
+      .compile(new Component {
+        val issue = new Dut()
+      })
+      .doSim { dut =>
+        dut.clockDomain.forkStimulus(10)
+        dut.clockDomain.waitSampling()
+      }
   }
 }

--- a/tester/src/test/scala/spinal/tester/scalatest/Issue963Tester.scala
+++ b/tester/src/test/scala/spinal/tester/scalatest/Issue963Tester.scala
@@ -1,0 +1,51 @@
+package spinal.tester.scalatest
+
+import org.scalatest.funsuite.AnyFunSuite
+
+import spinal.core._
+import spinal.lib._
+import spinal.core.sim._
+
+// unit test for issue #963
+
+// [info] - Minimal example for Issue963 which fails Spinal elaboration *** FAILED ***
+// [info]   java.lang.AssertionError: assertion failed
+// [info]   at scala.Predef$.assert(Predef.scala:156)
+// [info]   at spinal.core.package$.assert(core.scala:463)
+// [info]   at spinal.lib.whenIndexed$.apply(Utils.scala:1320)
+// [info]   at spinal.lib.StreamFragmentWidthAdapter$$anon$56$$anon$41$$anonfun$41.apply$mcV$sp(Stream.scala:1601)
+// [info]   at spinal.core.when$.apply(when.scala:90)
+// [info]   at spinal.lib.StreamFragmentWidthAdapter$$anon$56$$anon$41.<init>(Stream.scala:1600)
+// [info]   at spinal.lib.StreamFragmentWidthAdapter$$anon$56.<init>(Stream.scala:1562)
+// [info]   at spinal.lib.StreamFragmentWidthAdapter$.apply(Stream.scala:1541)
+// [info]   at spinal.tester.scalatest.Issue963$$anonfun$1$Dut$1.<init>(Issue963.scala:40)
+// [info]   at spinal.tester.scalatest.Issue963$$anonfun$1$$anonfun$apply$mcV$sp$1$$anon$2.<init>(Issue963.scala:45)
+// [info]   ...
+// [info] Run completed in 532 milliseconds.
+// [info] Total number of tests run: 1
+// [info] Suites: completed 1, aborted 0
+// [info] Tests: succeeded 0, failed 1, canceled 0, ignored 0, pending 0
+// [info] *** 1 TEST FAILED ***
+
+class Issue963Tester extends AnyFunSuite {
+  test("Minimal example for Issue963 which fails Spinal elaboration") {
+    class Dut extends Component {
+      val io = new Bundle {
+          val sink = slave Stream(Fragment(Bits(8 bits)))
+          val source = master Stream(Fragment(Bits(16 bits)))
+      }
+      val x = Stream(Fragment(Bits(16 bits)))
+
+      StreamFragmentWidthAdapter(io.sink, x, earlyLast = true)
+      io.source <-< x
+    }
+
+    SimConfig.withConfig(SpinalConfig(defaultClockDomainFrequency = FixedFrequency(4000 Hz))).compile(new Component{
+      val issue = new Dut()
+    })
+   .doSim{dut =>
+      dut.clockDomain.forkStimulus(10)
+      dut.clockDomain.waitSampling()
+    }
+  }
+}

